functions/v2/ocr/app/index.js

```diff
@@ -0,0 +1,223 @@
+ * @param {string} filename Cloud Storage file name.
+ * @returns {Promise}
+ */
+const detectText = async (bucketName, filename) => {
+  console.log(`Looking for text in image ${filename}`);
+  const [textDetections] = await vision.textDetection(
+    `gs://${bucketName}/${filename}`
+  );
+  const [annotation] = textDetections.textAnnotations;
+  const text = annotation ? annotation.description.trim() : '';
+  console.log('Extracted text from image:', text);
+
+  let [translateDetection] = await translate.detect(text);
+  if (Array.isArray(translateDetection)) {
+    [translateDetection] = translateDetection;
+  }
+  console.log(
+    `Detected language "${translateDetection.language}" for ${filename}`
+  );
+
+  // Submit a message to the bus for each language we're going to translate to
+  const TO_LANGS = process.env.TO_LANG.split(',');
+  const topicName = process.env.TRANSLATE_TOPIC;
+
+  const tasks = TO_LANGS.map(lang => {
+    const messageData = {
+      text: text,
+      filename: filename,
+      lang: lang,
+    };
+
+    // Helper function that publishes translation result to a Pub/Sub topic
+    // For more information on publishing Pub/Sub messages, see this page:
+    //   https://cloud.google.com/pubsub/docs/publisher
+    return publishResult(topicName, messageData);
+  });
+
+  return Promise.all(tasks);
+};
+// [END functions_ocr_detect]
+
+/**
+ * Appends a .txt suffix to the image name.
+ *
+ * @param {string} filename Name of a file.
+ * @param {string} lang Language to append.
+ * @returns {string} The new filename.
+ */
+const renameImageForSave = (filename, lang) => {
+  return `${filename}_to_${lang}.txt`;
+};
+
+// [START functions_ocr_process]
+/**
+ * This function is exported by index.js, and is executed when
+ * a file is uploaded to the Cloud Storage bucket you created
+ * for uploading images.
+ *
+ * @param {object} cloudEvent A CloudEvent containing the Cloud Storage File object.
+ */
+functions.cloudEvent('processImage', async cloudEvent => {
+  const {bucket, name} = cloudEvent.data;
+
+  if (!bucket) {
+    throw new Error(
+      'Bucket not provided. Make sure you have a "bucket" property in your request'
+    );
+  }
+  if (!name) {
+    throw new Error(
+      'Filename not provided. Make sure you have a "name" property in your request'
+    );
+  }
+
+  const results = await detectText(bucket, name);
+  console.log(`File ${name} processed.`);
+  return results;
+});
+// [END functions_ocr_process]
+
+// [START functions_ocr_translate]
+/**
+ * This function is exported by index.js, and is executed when
+ * a message is published to the Cloud Pub/Sub topic specified
+ * by the TRANSLATE_TOPIC environment variable. The function
+ * translates text using the Google Translate API.
+ *
+ * @param {object} cloudEvent The CloudEvent containing the Pub/Sub Message object
+ */
+functions.cloudEvent('translateText', async cloudEvent => {
+  const pubsubData = cloudEvent.data;
```

Comment:
nit: might be nice to link to https://cloud.google.com/storage/docs/json_api/v1/objects

Reply:
Added, thanks.